### ISSUE TITLE
Canvas area exceeds the maximum limit (width * heihgt > 16777216)

### ISSUE THREAD

**Issue by MianAzhar** [stephanrauh/ngx-extended-pdf-viewer#429]
I am using ngx-extended-pdf-viewer in Ionic Cordova project and it is working great in Android. But in iOS pdf is not displaying and it shows only white area and top controls and I get below error in console: 

`Canvas area exceeds the maximum limit (width * heihgt > 16777216)`

Can you please help me fix this this issue.

Thanks in advance!

**Comment by stephanrauh** [comment]
I can't. At least not quickly. You seem to have a PDF file with an exceptionally high resolution.

The error message says it all: every browser limits the maximum size of a canvas element. In the case of the Safari browser, that's 16777216 pixels. Reduce the resolution of your PDF file, and you're fine.

BTW, the same error message occurs on Mozilla's pdf.js showcase.

**Comment by MianAzhar** [comment]
@stephanrauh so currently there is no workaround for this? Any temporary hack for this?

**Comment by stephanrauh** [comment]
Wait, that's weird. Which version of ngx-extended-pdf-viewer are you using? A month ago, I've added a feature reduzing the canvas size on iOS. At the time, I was mostly interested in limiting the maximum zoom level, but it should work in your case, too. The error message you're reporting should be a thing of the past, at least with version 4.1.0+.

**Comment by MianAzhar** [comment]
I am using the latest version that is 4.1.0 and getting above error.

**Comment by stephanrauh** [comment]
That's unexpected. What happens when you open the PDF file with a low initial zoom? Something like `[zoom]="25%"`?

**Comment by stephanrauh** [comment]
Another idea: I'd like to know if your PDF file is the culprit, or if the application (or my library) is broken. Can you please check if your Cordova app copes with this file? https://pdfviewer.net/assets/pdfs/pdf-sample.pdf

**Comment by MianAzhar** [comment]
So I checked my pdf on https://pdfviewer.net/simple and it does not load there too in Safari iOS. If you say, I can share my pdf with you to check.

**Comment by stephanrauh** [comment]
Yes, that might be a good idea. Just drop me an e-mail (webmaster1 at beyondjava.de). You can also check the file at https://mozilla.github.io/pdf.js/es5/web/viewer.html. That's the showcase of the base library. I guess you can reproduce the bug over there, too.

**Comment by MianAzhar** [comment]
I just sent you PDF via email. 
Also got same issue on https://mozilla.github.io/pdf.js/es5/web/viewer.html

**Comment by stephanrauh** [comment]
Thank for sending me the file! I've managed to reproduce the bug. The file doesn't show on my iPad, either.

The file really seems to have a high resolution. When I open the file on my laptop, it crosses the 16 megapixels threshold when I zoom it to 110%. Maybe the sub-pixel resolution of the iOS devices increases the resolution by factor 4.

**Comment by stephanrauh** [comment]
I've found the problem. It's what I suspected, just a bit different. That's bad news because I can't help you.

What's happening is this: you PDF file has an embedded image with width = 7016 and height = 4964 pixels. The PDF viewer draws this image onto a canvas of the same size. It can't use a smaller canvas because that embedded image is provided as binary data. The size the PDF author chose is the size pdf.js uses, too. After that, pdf.js scales the image to according to your `[zoom]` setting - but the error occurs earlier.

@timvandermeij Maybe you've got an idea. Is it possible to modify pdf.js to render pdf files with embedded images exceeding the browser's limits? I'd even offer to provide a pull request.
